File: packages/study_logic/src/study_logic/vault.py
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from collections.abc import Callable, Mapping, Sequence
from typing import Any, Protocol

from study_logic.models import DEFAULT_SAMPLE_LIMIT, DEFAULT_TOP_K, Chunk, is_citable

RetrieveFn = Callable[..., Sequence[Chunk | Mapping[str, Any]]]
ListChunksFn = Callable[..., Sequence[Chunk | Mapping[str, Any]]]


class VaultRetrieve(Protocol):
    def retrieve(self, notebook_id: str, query: str, top_k: int = DEFAULT_TOP_K) -> Sequence[Chunk | Mapping[str, Any]]:
        ...

    def list_chunks(
        self, notebook_id: str, limit: int = DEFAULT_SAMPLE_LIMIT
    ) -> Sequence[Chunk | Mapping[str, Any]]:
        ...


def normalize_chunks(raw: Sequence[Chunk | Mapping[str, Any]]) -> list[Chunk]:
    out: list[Chunk] = []
    for item in raw:
        chunk = item if isinstance(item, Chunk) else Chunk.from_mapping(item)
        if is_citable(chunk):
            out.append(chunk)
    return out


def call_retrieve(retrieve: RetrieveFn, notebook_id: str, query: str, top_k: int = DEFAULT_TOP_K) -> list[Chunk]:
    result = retrieve(notebook_id, query, top_k)
    return normalize_chunks(list(result))


def call_list_chunks(
    list_chunks: ListChunksFn, notebook_id: str, limit: int = DEFAULT_SAMPLE_LIMIT
) -> list[Chunk]:
    result = list_chunks(notebook_id, limit)
    return normalize_chunks(list(result))


def companion_list_chunks(retrieve: RetrieveFn | None) -> ListChunksFn | None:
    """If retrieve is a bound VaultRetrieve.retrieve, reuse the same object's list_chunks."""
    owner = getattr(retrieve, "__self__", None)
    companion = getattr(owner, "list_chunks", None)
    return companion if callable(companion) else None


class InMemoryVault:
    def __init__(self, seed: Mapping[str, Sequence[Chunk | Mapping[str, Any]]] | None = None) -> None:
        self._chunks: dict[str, list[Chunk]] = {}
        if seed:
            for notebook_id, chunks in seed.items():
                self._chunks[notebook_id] = [c if isinstance(c, Chunk) else Chunk.from_mapping(c) for c in chunks]

    def retrieve(self, notebook_id: str, query: str, top_k: int = DEFAULT_TOP_K) -> list[Chunk]:
        all_chunks = [c for c in self._chunks.get(notebook_id, []) if is_citable(c)]
        q = query.strip().lower()
        if not q:
            return [Chunk(**{**c.__dict__, "score": c.score or 1}) for c in all_chunks[:top_k]]
        terms = [t for t in q.split() if t]
        ranked: list[tuple[float, Chunk]] = []
        for chunk in all_chunks:
            hay = f"{chunk.text} {chunk.source_filename} {chunk.source_id} {chunk.locator}".lower()
            score = 0.0
            if q in hay:
                score += 5
            score += sum(1 for t in terms if t in hay)
            if score > 0:
                ranked.append((score, chunk))
        ranked.sort(key=lambda row: row[0], reverse=True)
        return [Chunk(**{**chunk.__dict__, "score": score}) for score, chunk in ranked[:top_k]]

    def list_chunks(self, notebook_id: str, limit: int = DEFAULT_SAMPLE_LIMIT) -> list[Chunk]:
        all_chunks = [c for c in self._chunks.get(notebook_id, []) if is_citable(c)]
        return [Chunk(**{**c.__dict__, "score": c.score or 1}) for c in all_chunks[:limit]]

    def get_chunk(self, chunk_id: str) -> Chunk | None:
        for chunks in self._chunks.values():
            for chunk in chunks:
                if chunk.id == chunk_id:
                    return Chunk(**chunk.__dict__)
        return None


class HttpVaultRetrieve:
    """HTTP client mode: POST retrieve for quiz; GET …/chunks for propose."""

    def __init__(self, base_url: str) -> None:
        self.base_url = base_url.rstrip("/")

    def retrieve(self, notebook_id: str, query: str, top_k: int = DEFAULT_TOP_K) -> list[Chunk]:
        url = f"{self.base_url}/api/v1/notebooks/{notebook_id}/retrieve"
        payload = json.dumps({"query": query, "top_k": top_k}).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=payload,
            headers={"content-type": "application/json"},
            method="POST",
        )
        try:
            with urllib.request.urlopen(req) as res:
                body = json.loads(res.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise RuntimeError(f"Vault retrieve failed: {exc.code}") from exc
        chunks = body.get("chunks") if isinstance(body, dict) else body
        if not isinstance(chunks, list):
            return []
        return normalize_chunks(chunks)

    def list_chunks(self, notebook_id: str, limit: int = DEFAULT_SAMPLE_LIMIT) -> list[Chunk]:
        encoded = urllib.parse.quote(notebook_id, safe="")
        query = urllib.parse.urlencode({"limit": limit})
        url = f"{self.base_url}/api/v1/notebooks/{encoded}/chunks?{query}"
        req = urllib.request.Request(url, method="GET")
        try:
            with urllib.request.urlopen(req) as res:
                body = json.loads(res.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raise RuntimeError(f"Vault list_chunks failed: {exc.code}") from exc
        if isinstance(body, list):
            chunks = body
        elif isinstance(body, dict):
            chunks = body.get("chunks")
        else:
            chunks = None
        if not isinstance(chunks, list):
            return []
        return normalize_chunks(chunks)


FIXTURE_NOTEBOOK_ID = "nb_bio"


def fixture_chunks() -> list[Chunk]:
    return [
        Chunk(
            id="chunk_mitosis",
            source_id="notes-cell-cycle",
            text=(
                "# Mitosis\n"
                "Mitosis is a type of cell division that produces two genetically identical daughter cells. "
                "The stages of mitosis are prophase, metaphase, anaphase, and telophase. "
                "During metaphase, chromosomes align at the cell equator."
            ),
            locator="notes-cell-cycle.md#mitosis",
            score=1,
            source_filename="notes-cell-cycle.md",
        ),
        Chunk(
            id="chunk_meiosis",
            source_id="notes-cell-cycle",
            text=(
                "# Meiosis\n"
                "Meiosis produces four haploid gametes and includes two rounds of division. "
                "Crossing over occurs during prophase I and increases genetic variation."
            ),
            locator="notes-cell-cycle.md#meiosis",
            score=1,
            source_filename="notes-cell-cycle.md",
        ),
        Chunk(
            id="chunk_photosynthesis",
            source_id="notes-energy",
            text=(
                "# Photosynthesis\n"
                "Photosynthesis converts light energy into chemical energy stored in sugars. "
                "The light-dependent reactions occur in the thylakoid membrane and produce ATP and NADPH."
            ),
            locator="notes-energy.md#photosynthesis",
            score=1,
            source_filename="notes-energy.md",
        ),
    ]


def create_fixture_vault() -> InMemoryVault:
    return InMemoryVault({FIXTURE_NOTEBOOK_ID: fixture_chunks()})

File: packages/study-logic/src/vault.ts
import type { Chunk, RetrieveArgs, VaultRetrieve } from "./types.js";

export function isCitableChunk(chunk: Chunk): boolean {
  return Boolean(chunk.id?.trim()) && Boolean(chunk.text?.trim());
}

/**
 * In-memory fixture vault for tests and local smoke.
 * Backend should implement VaultRetrieve against:
 *   POST /notebooks/:id/retrieve  { query, top_k? } → Chunk[]
 *   GET  /chunks/:id              → Chunk
 */
export class InMemoryVault implements VaultRetrieve {
  private readonly chunks = new Map<string, Chunk[]>();

  constructor(seed?: Record<string, Chunk[]>) {
    if (seed) {
      for (const [notebookId, list] of Object.entries(seed)) {
        this.chunks.set(notebookId, list.map((c) => ({ ...c })));
      }
    }
  }

  seed(notebookId: string, chunks: Chunk[]): void {
    this.chunks.set(notebookId, chunks.map((c) => ({ ...c })));
  }

  getChunk(id: string): Chunk | undefined {
    for (const list of this.chunks.values()) {
      const found = list.find((c) => c.id === id);
      if (found) return { ...found };
    }
    return undefined;
  }

  allChunks(notebookId: string): Chunk[] {
    return (this.chunks.get(notebookId) ?? []).map((c) => ({ ...c }));
  }

  async retrieve(args: RetrieveArgs): Promise<Chunk[]> {
    const topK = args.top_k ?? 8;
    const all = this.chunks.get(args.notebook_id) ?? [];
    const query = args.query.trim().toLowerCase();
    if (!query) {
      return all.filter(isCitableChunk).slice(0, topK).map((c) => ({ ...c }));
    }

    const terms = query.split(/\s+/).filter(Boolean);
    const ranked = all
      .filter(isCitableChunk)
      .map((chunk) => ({ chunk, score: scoreChunk(chunk, query, terms) }))
      .filter((row) => row.score > 0)
      .sort((a, b) => b.score - a.score)
      .slice(0, topK)
      .map((row) => ({ ...row.chunk }));

    return ranked;
  }
}

function scoreChunk(chunk: Chunk, query: string, terms: string[]): number {
  const hay = `${chunk.text} ${chunk.source_label ?? ""} ${chunk.source_id ?? ""}`.toLowerCase();
  let score = 0;
  if (hay.includes(query)) score += 5;
  for (const term of terms) {
    if (hay.includes(term)) score += 1;
  }
  return score;
}

/**
 * HTTP adapter for a real Backend vault. Study-logic consumes these routes;
 * it does not own embeddings or OCR.
 */
export class HttpVaultRetrieve implements VaultRetrieve {
  constructor(private readonly baseUrl: string) {}

  async retrieve(args: RetrieveArgs): Promise<Chunk[]> {
    const url = `${trimSlash(this.baseUrl)}/notebooks/${encodeURIComponent(args.notebook_id)}/retrieve`;
    const res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ query: args.query, top_k: args.top_k }),
    });
    if (!res.ok) {
      throw new Error(`Vault retrieve failed: ${res.status}`);
    }
    const body = (await res.json()) as { chunks?: Chunk[] } | Chunk[];
    return Array.isArray(body) ? body : (body.chunks ?? []);
  }

  async getChunk(id: string): Promise<Chunk | undefined> {
    const url = `${trimSlash(this.baseUrl)}/chunks/${encodeURIComponent(id)}`;
    const res = await fetch(url);
    if (res.status === 404) return undefined;
    if (!res.ok) {
      throw new Error(`Vault getChunk failed: ${res.status}`);
    }
    return (await res.json()) as Chunk;
  }
}

function trimSlash(url: string): string {
  return url.replace(/\/+$/, "");
}

export const FIXTURE_NOTEBOOK_ID = "nb_bio";

export function fixtureChunks(): Chunk[] {
  return [
    {
      id: "chunk_mitosis",
      source_id: "notes-cell-cycle",
      source_label: "Mitosis",
      text:
        "# Mitosis\n" +
        "Mitosis is a type of cell division that produces two genetically identical daughter cells. " +
        "The stages of mitosis are prophase, metaphase, anaphase, and telophase. " +
        "During metaphase, chromosomes align at the cell equator.",
    },
    {
      id: "chunk_meiosis",
      source_id: "notes-cell-cycle",
      source_label: "Meiosis",
      text:
        "# Meiosis\n" +
        "Meiosis produces four haploid gametes and includes two rounds of division. " +
        "Crossing over occurs during prophase I and increases genetic variation.",
    },
    {
      id: "chunk_photosynthesis",
      source_id: "notes-energy",
      source_label: "Photosynthesis",
      text:
        "# Photosynthesis\n" +
        "Photosynthesis converts light energy into chemical energy stored in sugars. " +
        "The light-dependent reactions occur in the thylakoid membrane and produce ATP and NADPH.",
    },
  ];
}

export function createFixtureVault(): InMemoryVault {
  return new InMemoryVault({ [FIXTURE_NOTEBOOK_ID]: fixtureChunks() });
}
